enhancements/upgrade-bundle.md

```diff
@@ -0,0 +1,259 @@
+This enhancement will add a new subcommand, `bundle-upgrade`,
+to the `run` command. This new command, `run bundle-upgrade`, will perform the upgrade of an operator from a previous version to a new specified version.
+
+Currently, the run command has different subcommands: 
+
+ * `run local`
+ * `run packagemanifests`
+ * `run bundle` - currently being implemented, design document is covered by [run bundle enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/run-bundle.md)
+
+For doing an `bundle-upgrade`, there could be two scenarios: 
+
+* Case 1 - A previous version of the operator is already running on the cluster, and the index image contains this operator version. Now, we we want to upgrade the in cluster running operator to a new version that we specify. 
+
+* Case 2 - No previous version of an operator is running on the cluster, so we provide the new bundle image which is accessible and pullable within the cluster, which contains our desired version of the operator for the upgrade. 
+
+In case 1, we will have to figure out which catalog source the existing operator was installed in and somehow dynamically inject the new bundle image into the index image for OLM to detect this change and upgrade to the new version we provided. 
+
+In this proposal, we will design bundle upgrade for case 2. For this case, we also assume that an index image already exists that has the from version of the operator bundle. So we don't create a new index image. We use the existing index image (which is mandatory to provide) for the operator and provide the new bundle image for the upgrade. For this to happen, we need to specify starting CSV.
+Generally, when there is no starting CSV specified, OLM always detects and installs the latest version available for an operator in a particular  channel. For example, in production case, if an operator is running and there's a subscription for it, and there comes a fix available in a stable channel, OLM will detect and upgrade by default. 
+
+Let's consider a few scenarios: 
+
+How do we verfiy whether an operator has been upgraded successfully?
+To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully installed or upgraded. 
+  
+In case of subscription failure, the end user needs to be aware of this. If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. 
```

Comment:
This gets back to my question above about whether we need to know the version we're upgrading to. For us to verify that the upgrade ends with the bundle provided on the command line, I agree that this seems like a good approach.

There is one possible gotcha though: how do we know if OLM is done upgrading? Here's a scenario:
* Starting CSV: `A`
* Bundle provided on command line: `B`
* Upgrade graph of existing index image: `A -> C`
* Upgrade graph of ephemeral index after inserting B: `A -> B -> C`

After OLM upgrades A to B, our check would say the upgrade was successful, but then simultaneously OLM is probably gearing up for the upgrade from `B` to `C`. Our command then returns saying we successfully upgraded, but then the user goes and looks at his cluster and sees that `C` is installed when they expected `B`.

Technically, OLM is never "finished" making upgrades, right? The entire idea is that if a new operator shows up in the upgrade graph, OLM will upgrade it.

@shawn-hurley @ecordell For our purposes, is there a reasonable way to know when OLM has made its way though the upgrade graph of the channel in our ephemeral index?

Another similar scenario is:
* Starting CSV: `A`
* Bundle provided on command line: `Z`
* Upgrade graph of existing index image: `A -> B -> C -> ... -> X -> Y`
* Upgrade graph of ephemeral index after inserting Z: `A -> B -> C -> ... -> X -> Y -> Z`

This would take OLM awhile to roll through right?

Is there a way for SDK to know the upgrade graph for a channel starting with `--starting-csv` and ending with the HEAD of the channel?

Reply:
Scenario 1: all upgrades up to `B` were successful, and that's all the command guarantees to report right? If `run bundle-upgrade` tries to report on operator versions "later" in the upgrade graph then semantically the command is always checking the entire upgrade channel and not one specific upgrade path.

Scenario 2: the naive approach would be to poll for operator version `Z` until a timeout ends or `Z` installs/fails to install. One way I see around having to tune the timeout based on how many operators are in an upgrade graph is, like you suggest, to rebuild the graph and poll for each operator in it until we reach `Z`, resetting the timeout for each correct installation. I'm sure there's a way to get this info from the index db.